From 7c925c45022887504a83f9760e79a192ec5f52e7 Mon Sep 17 00:00:00 2001
From: Amelia Crate <amelia.crate@chainguard.dev>
Date: Thu, 4 Dec 2025 15:17:33 -0600
Subject: [PATCH 2/4] crypto: reusable jitterentropy

Vendor sha3 for jitterentropy usage. In 6.12 we did this by managing SHA3 state
and calling sha3 functions directly but the SHA3 algorithm no longer implements
partial block handling but relies on the crypto API level partial block handling,
so this is no longer feasible.

* copy sha3_generic to sha3-jitterentropy
* shim function calls in certified sha3-jitterentropy
* move crypto algorithm registration and macros to a recompilable sha3-jitterentropy-kcapi object
* register algorithm on jitterentropy-kcapi startup
---
 crypto/jitterentropy-kcapi.c      |  10 +-
 crypto/jitterentropy-shim.c       |  10 ++
 crypto/jitterentropy-shim.h       |   2 +
 crypto/sha3-jitterentropy-kcapi.c |  43 ++++++
 crypto/sha3-jitterentropy.c       | 217 ++++++++++++++++++++++++++++++
 crypto/sha3_jitter.h              |  16 +++
 include/crypto/sha3_jitter.h      |  38 ++++++
 7 files changed, 334 insertions(+), 2 deletions(-)
 create mode 100644 crypto/sha3-jitterentropy-kcapi.c
 create mode 100644 crypto/sha3-jitterentropy.c
 create mode 100644 crypto/sha3_jitter.h
 create mode 100644 include/crypto/sha3_jitter.h

diff --git a/crypto/jitterentropy-kcapi.c b/crypto/jitterentropy-kcapi.c
index 064581fedf707..6f5e8f164b630 100644
--- a/crypto/jitterentropy-kcapi.c
+++ b/crypto/jitterentropy-kcapi.c
@@ -37,7 +37,7 @@
  * DAMAGE.
  */
 #include <crypto/hash.h>
-#include <crypto/sha3.h>
+#include <crypto/sha3_jitter.h>
 #include <linux/fips.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
@@ -47,7 +47,7 @@
 
 #include "jitterentropy.h"
 
-#define JENT_CONDITIONING_HASH	"sha3-256-generic"
+#define JENT_CONDITIONING_HASH	"sha3-256-jitter"
 
 /***************************************************************************
  * Helper function
@@ -250,6 +250,11 @@ static int __init jent_mod_init(void)
 	struct crypto_shash *tfm;
 	int ret = 0;
 
+	ret = sha3_jitter_mod_init();
+	if (ret) {
+		return ret;
+	}
+
 	jent_testing_init();
 
 	tfm = crypto_alloc_shash(JENT_CONDITIONING_HASH, 0, 0);
@@ -277,6 +282,7 @@ static int __init jent_mod_init(void)
 
 static void __exit jent_mod_exit(void)
 {
+	sha3_jitter_mod_exit();
 	jent_testing_exit();
 	crypto_unregister_rng(&jent_alg);
 }
diff --git a/crypto/jitterentropy-shim.c b/crypto/jitterentropy-shim.c
index 2cf6c9b3fe211..497e0df7b26e3 100644
--- a/crypto/jitterentropy-shim.c
+++ b/crypto/jitterentropy-shim.c
@@ -359,3 +359,13 @@ unsigned int jent_shim_crypto_shash_digestsize(void *tfm)
 {
 	return crypto_shash_digestsize((struct crypto_shash *)tfm);
 }
+
+void *jent_shim_shash_desc_ctx(void *desc)
+{
+	return shash_desc_ctx((struct shash_desc *)desc);
+}
+
+unsigned int jent_shim_crypto_shash_blocksize(void *tfm)
+{
+	return crypto_shash_blocksize((struct crypto_shash *)tfm);
+}
diff --git a/crypto/jitterentropy-shim.h b/crypto/jitterentropy-shim.h
index 5b957c31871cb..9f9720ec13813 100644
--- a/crypto/jitterentropy-shim.h
+++ b/crypto/jitterentropy-shim.h
@@ -111,5 +111,7 @@ int jent_shim_crypto_shash_update(void *desc, const void *data, unsigned int len
 int jent_shim_crypto_shash_finup(void *desc, const void *data, unsigned int len, void *out);
 int jent_shim_crypto_shash_final(void *desc, void *out);
 unsigned int jent_shim_crypto_shash_digestsize(void *tfm);
+void *jent_shim_shash_desc_ctx(void *desc);
+unsigned int jent_shim_crypto_shash_blocksize(void *tfm);
 
 #endif /* _JITTERENTROPY_SHIM_H */
diff --git a/crypto/sha3-jitterentropy-kcapi.c b/crypto/sha3-jitterentropy-kcapi.c
new file mode 100644
index 0000000000000..4730431d5ce35
--- /dev/null
+++ b/crypto/sha3-jitterentropy-kcapi.c
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Cryptographic API.
+ *
+ * SHA-3, as specified in
+ * https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf
+ *
+ * SHA-3 code by Jeff Garzik <jeff@garzik.org>
+ *               Ard Biesheuvel <ard.biesheuvel@linaro.org>
+ */
+#include <crypto/internal/hash.h>
+#include <crypto/sha3_jitter.h>
+#include "sha3_jitter.h"
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/string.h>
+#include <linux/unaligned.h>
+
+EXPORT_SYMBOL(crypto_sha3_jitter_init);
+
+static struct shash_alg algs[] = { {
+	.digestsize		= SHA3_256_DIGEST_SIZE,
+	.init			= crypto_sha3_init,
+	.update			= crypto_sha3_update,
+	.finup			= crypto_sha3_finup,
+	.descsize		= SHA3_STATE_SIZE,
+	.base.cra_name		= "sha3-256-jitter",
+	.base.cra_driver_name	= "sha3-256-jitter",
+	.base.cra_flags		= CRYPTO_AHASH_ALG_BLOCK_ONLY,
+	.base.cra_blocksize	= SHA3_256_BLOCK_SIZE,
+	.base.cra_module	= THIS_MODULE,
+} };
+
+int sha3_jitter_mod_init(void)
+{
+	return crypto_register_shashes(algs, ARRAY_SIZE(algs));
+}
+
+void sha3_jitter_mod_exit(void)
+{
+	crypto_unregister_shashes(algs, ARRAY_SIZE(algs));
+}
+
diff --git a/crypto/sha3-jitterentropy.c b/crypto/sha3-jitterentropy.c
new file mode 100644
index 0000000000000..66da4fca33f80
--- /dev/null
+++ b/crypto/sha3-jitterentropy.c
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Cryptographic API.
+ *
+ * SHA-3, as specified in
+ * https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf
+ *
+ * SHA-3 code by Jeff Garzik <jeff@garzik.org>
+ *               Ard Biesheuvel <ard.biesheuvel@linaro.org>
+ */
+#include <crypto/sha3_jitter.h>
+#include <crypto/internal/hash.h>
+#include "sha3_jitter.h"
+#include "jitterentropy-shim.h"
+
+/*
+ * On some 32-bit architectures (h8300), GCC ends up using
+ * over 1 KB of stack if we inline the round calculation into the loop
+ * in keccakf(). On the other hand, on 64-bit architectures with plenty
+ * of [64-bit wide] general purpose registers, not inlining it severely
+ * hurts performance. So let's use 64-bitness as a heuristic to decide
+ * whether to inline or not.
+ */
+#ifdef CONFIG_64BIT
+#define SHA3_INLINE	inline
+#else
+#define SHA3_INLINE	noinline
+#endif
+
+#define KECCAK_ROUNDS 24
+
+static const u64 keccakf_rndc[24] = {
+	0x0000000000000001ULL, 0x0000000000008082ULL, 0x800000000000808aULL,
+	0x8000000080008000ULL, 0x000000000000808bULL, 0x0000000080000001ULL,
+	0x8000000080008081ULL, 0x8000000000008009ULL, 0x000000000000008aULL,
+	0x0000000000000088ULL, 0x0000000080008009ULL, 0x000000008000000aULL,
+	0x000000008000808bULL, 0x800000000000008bULL, 0x8000000000008089ULL,
+	0x8000000000008003ULL, 0x8000000000008002ULL, 0x8000000000000080ULL,
+	0x000000000000800aULL, 0x800000008000000aULL, 0x8000000080008081ULL,
+	0x8000000000008080ULL, 0x0000000080000001ULL, 0x8000000080008008ULL
+};
+
+/* update the state with given number of rounds */
+
+SHA3_INLINE void keccakf_round(u64 st[25])
+{
+	u64 t[5], tt, bc[5];
+
+	/* Theta */
+	bc[0] = st[0] ^ st[5] ^ st[10] ^ st[15] ^ st[20];
+	bc[1] = st[1] ^ st[6] ^ st[11] ^ st[16] ^ st[21];
+	bc[2] = st[2] ^ st[7] ^ st[12] ^ st[17] ^ st[22];
+	bc[3] = st[3] ^ st[8] ^ st[13] ^ st[18] ^ st[23];
+	bc[4] = st[4] ^ st[9] ^ st[14] ^ st[19] ^ st[24];
+
+	t[0] = bc[4] ^ jent_shim_rol64(bc[1], 1);
+	t[1] = bc[0] ^ jent_shim_rol64(bc[2], 1);
+	t[2] = bc[1] ^ jent_shim_rol64(bc[3], 1);
+	t[3] = bc[2] ^ jent_shim_rol64(bc[4], 1);
+	t[4] = bc[3] ^ jent_shim_rol64(bc[0], 1);
+
+	st[0] ^= t[0];
+
+	/* Rho Pi */
+	tt = st[1];
+	st[ 1] = jent_shim_rol64(st[ 6] ^ t[1], 44);
+	st[ 6] = jent_shim_rol64(st[ 9] ^ t[4], 20);
+	st[ 9] = jent_shim_rol64(st[22] ^ t[2], 61);
+	st[22] = jent_shim_rol64(st[14] ^ t[4], 39);
+	st[14] = jent_shim_rol64(st[20] ^ t[0], 18);
+	st[20] = jent_shim_rol64(st[ 2] ^ t[2], 62);
+	st[ 2] = jent_shim_rol64(st[12] ^ t[2], 43);
+	st[12] = jent_shim_rol64(st[13] ^ t[3], 25);
+	st[13] = jent_shim_rol64(st[19] ^ t[4],  8);
+	st[19] = jent_shim_rol64(st[23] ^ t[3], 56);
+	st[23] = jent_shim_rol64(st[15] ^ t[0], 41);
+	st[15] = jent_shim_rol64(st[ 4] ^ t[4], 27);
+	st[ 4] = jent_shim_rol64(st[24] ^ t[4], 14);
+	st[24] = jent_shim_rol64(st[21] ^ t[1],  2);
+	st[21] = jent_shim_rol64(st[ 8] ^ t[3], 55);
+	st[ 8] = jent_shim_rol64(st[16] ^ t[1], 45);
+	st[16] = jent_shim_rol64(st[ 5] ^ t[0], 36);
+	st[ 5] = jent_shim_rol64(st[ 3] ^ t[3], 28);
+	st[ 3] = jent_shim_rol64(st[18] ^ t[3], 21);
+	st[18] = jent_shim_rol64(st[17] ^ t[2], 15);
+	st[17] = jent_shim_rol64(st[11] ^ t[1], 10);
+	st[11] = jent_shim_rol64(st[ 7] ^ t[2],  6);
+	st[ 7] = jent_shim_rol64(st[10] ^ t[0],  3);
+	st[10] = jent_shim_rol64(    tt ^ t[1],  1);
+
+	/* Chi */
+	bc[ 0] = ~st[ 1] & st[ 2];
+	bc[ 1] = ~st[ 2] & st[ 3];
+	bc[ 2] = ~st[ 3] & st[ 4];
+	bc[ 3] = ~st[ 4] & st[ 0];
+	bc[ 4] = ~st[ 0] & st[ 1];
+	st[ 0] ^= bc[ 0];
+	st[ 1] ^= bc[ 1];
+	st[ 2] ^= bc[ 2];
+	st[ 3] ^= bc[ 3];
+	st[ 4] ^= bc[ 4];
+
+	bc[ 0] = ~st[ 6] & st[ 7];
+	bc[ 1] = ~st[ 7] & st[ 8];
+	bc[ 2] = ~st[ 8] & st[ 9];
+	bc[ 3] = ~st[ 9] & st[ 5];
+	bc[ 4] = ~st[ 5] & st[ 6];
+	st[ 5] ^= bc[ 0];
+	st[ 6] ^= bc[ 1];
+	st[ 7] ^= bc[ 2];
+	st[ 8] ^= bc[ 3];
+	st[ 9] ^= bc[ 4];
+
+	bc[ 0] = ~st[11] & st[12];
+	bc[ 1] = ~st[12] & st[13];
+	bc[ 2] = ~st[13] & st[14];
+	bc[ 3] = ~st[14] & st[10];
+	bc[ 4] = ~st[10] & st[11];
+	st[10] ^= bc[ 0];
+	st[11] ^= bc[ 1];
+	st[12] ^= bc[ 2];
+	st[13] ^= bc[ 3];
+	st[14] ^= bc[ 4];
+
+	bc[ 0] = ~st[16] & st[17];
+	bc[ 1] = ~st[17] & st[18];
+	bc[ 2] = ~st[18] & st[19];
+	bc[ 3] = ~st[19] & st[15];
+	bc[ 4] = ~st[15] & st[16];
+	st[15] ^= bc[ 0];
+	st[16] ^= bc[ 1];
+	st[17] ^= bc[ 2];
+	st[18] ^= bc[ 3];
+	st[19] ^= bc[ 4];
+
+	bc[ 0] = ~st[21] & st[22];
+	bc[ 1] = ~st[22] & st[23];
+	bc[ 2] = ~st[23] & st[24];
+	bc[ 3] = ~st[24] & st[20];
+	bc[ 4] = ~st[20] & st[21];
+	st[20] ^= bc[ 0];
+	st[21] ^= bc[ 1];
+	st[22] ^= bc[ 2];
+	st[23] ^= bc[ 3];
+	st[24] ^= bc[ 4];
+}
+
+void keccakf(u64 st[25])
+{
+	int round;
+
+	for (round = 0; round < KECCAK_ROUNDS; round++) {
+		keccakf_round(st);
+		/* Iota */
+		st[0] ^= keccakf_rndc[round];
+	}
+}
+
+int crypto_sha3_jitter_init(struct shash_desc *desc)
+{
+	struct sha3_state *sctx = jent_shim_shash_desc_ctx(desc);
+
+	jent_shim_memset(sctx->st, 0, sizeof(sctx->st));
+	return 0;
+}
+
+int crypto_sha3_update(struct shash_desc *desc, const u8 *data,
+			      unsigned int len)
+{
+	unsigned int rsiz = jent_shim_crypto_shash_blocksize(desc->tfm);
+	struct sha3_state *sctx = jent_shim_shash_desc_ctx(desc);
+	unsigned int rsizw = rsiz / 8;
+
+	do {
+		int i;
+
+		for (i = 0; i < rsizw; i++)
+			sctx->st[i] ^= jent_shim_get_unaligned_le64(data + 8 * i);
+		keccakf(sctx->st);
+
+		data += rsiz;
+		len -= rsiz;
+	} while (len >= rsiz);
+	return len;
+}
+
+int crypto_sha3_finup(struct shash_desc *desc, const u8 *src,
+			     unsigned int len, u8 *out)
+{
+	unsigned int digest_size = jent_shim_crypto_shash_digestsize(desc->tfm);
+	unsigned int rsiz = jent_shim_crypto_shash_blocksize(desc->tfm);
+	struct sha3_state *sctx = jent_shim_shash_desc_ctx(desc);
+	__le64 block[SHA3_224_BLOCK_SIZE / 8] = {};
+	__le64 *digest = (__le64 *)out;
+	unsigned int rsizw = rsiz / 8;
+	u8 *p;
+	int i;
+
+	p = jent_shim_memcpy(block, src, len);
+	p[len++] = 0x06;
+	p[rsiz - 1] |= 0x80;
+
+	for (i = 0; i < rsizw; i++)
+		sctx->st[i] ^= le64_to_cpu(block[i]);
+	jent_shim_memzero_explicit(block, sizeof(block));
+
+	keccakf(sctx->st);
+
+	for (i = 0; i < digest_size / 8; i++)
+		jent_shim_put_unaligned_le64(sctx->st[i], digest++);
+
+	if (digest_size & 4)
+		jent_shim_put_unaligned_le32(sctx->st[i], (__le32 *)digest);
+
+	return 0;
+}
+
diff --git a/crypto/sha3_jitter.h b/crypto/sha3_jitter.h
new file mode 100644
index 0000000000000..59799df44e294
--- /dev/null
+++ b/crypto/sha3_jitter.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Internal functions for sha3 jitterentropy
+ */
+#ifndef __CRYPTO_SHA3_JITTER_INTERNAL_H__
+#define __CRYPTO_SHA3_JITTER_INTERNAL_H__
+
+void keccakf_round(u64 st[25]);
+void keccakf(u64 st[25]);
+int crypto_sha3_init(struct shash_desc *desc);
+int crypto_sha3_update(struct shash_desc *desc, const u8 *data,
+			      unsigned int len);
+int crypto_sha3_finup(struct shash_desc *desc, const u8 *src,
+			     unsigned int len, u8 *out);
+
+#endif
diff --git a/include/crypto/sha3_jitter.h b/include/crypto/sha3_jitter.h
new file mode 100644
index 0000000000000..c4d37731af5dc
--- /dev/null
+++ b/include/crypto/sha3_jitter.h
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Common values for SHA-3 algorithms
+ */
+#ifndef __CRYPTO_SHA3_JITTER_H__
+#define __CRYPTO_SHA3_JITTER_H__
+
+#include <linux/types.h>
+
+#define SHA3_224_DIGEST_SIZE	(224 / 8)
+#define SHA3_224_BLOCK_SIZE	(200 - 2 * SHA3_224_DIGEST_SIZE)
+#define SHA3_224_EXPORT_SIZE	SHA3_STATE_SIZE + SHA3_224_BLOCK_SIZE + 1
+
+#define SHA3_256_DIGEST_SIZE	(256 / 8)
+#define SHA3_256_BLOCK_SIZE	(200 - 2 * SHA3_256_DIGEST_SIZE)
+#define SHA3_256_EXPORT_SIZE	SHA3_STATE_SIZE + SHA3_256_BLOCK_SIZE + 1
+
+#define SHA3_384_DIGEST_SIZE	(384 / 8)
+#define SHA3_384_BLOCK_SIZE	(200 - 2 * SHA3_384_DIGEST_SIZE)
+#define SHA3_384_EXPORT_SIZE	SHA3_STATE_SIZE + SHA3_384_BLOCK_SIZE + 1
+
+#define SHA3_512_DIGEST_SIZE	(512 / 8)
+#define SHA3_512_BLOCK_SIZE	(200 - 2 * SHA3_512_DIGEST_SIZE)
+#define SHA3_512_EXPORT_SIZE	SHA3_STATE_SIZE + SHA3_512_BLOCK_SIZE + 1
+
+#define SHA3_STATE_SIZE		200
+
+struct shash_desc;
+
+struct sha3_state {
+	u64		st[SHA3_STATE_SIZE / 8];
+};
+
+int crypto_sha3_jitter_init(struct shash_desc *desc);
+
+int sha3_jitter_mod_init(void);
+void sha3_jitter_mod_exit(void);
+#endif
-- 
2.50.1

